File: ios-swiftui/library/Sources/PlanetComponents/TspUpload.swift
import SwiftUI

public struct TspUploadFile: Identifiable, Equatable {
    public var id: String
    public var name: String

    public init(id: String, name: String) {
        self.id = id
        self.name = name
    }
}

/// Lightweight file list + add trigger (mock add; no real picker backend required).
public struct TspUpload: View {
    public var files: [TspUploadFile]
    public var multiple: Bool
    public var disabled: Bool
    public var accept: String?
    public var theme: StarPlanetTheme
    public var onChange: ([TspUploadFile]) -> Void
    public var onRemove: (TspUploadFile) -> Void
    public var onAdd: (() -> [TspUploadFile])?

    public init(
        files: [TspUploadFile] = [],
        multiple: Bool = true,
        disabled: Bool = false,
        accept: String? = nil,
        theme: StarPlanetTheme = .sky,
        onChange: @escaping ([TspUploadFile]) -> Void = { _ in },
        onRemove: @escaping (TspUploadFile) -> Void = { _ in },
        onAdd: (() -> [TspUploadFile])? = nil
    ) {
        self.files = files
        self.multiple = multiple
        self.disabled = disabled
        self.accept = accept
        self.theme = theme
        self.onChange = onChange
        self.onRemove = onRemove
        self.onAdd = onAdd
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            Button(action: handleAdd) {
                Text("选择文件")
                    .font(.system(size: 14, weight: .heavy))
                    .foregroundColor(theme.brandPrimary)
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, 12)
                    .background(theme.brandSubtle)
                    .overlay(RoundedRectangle(cornerRadius: 16).stroke(theme.borderDefault, lineWidth: 2))
                    .clipShape(RoundedRectangle(cornerRadius: 16))
            }
            .buttonStyle(.plain)
            .disabled(disabled)

            if files.isEmpty {
                Text("尚未选择文件")
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundColor(theme.textTertiary)
            } else {
                ForEach(files) { file in
                    HStack {
                        Text(file.name)
                            .font(.system(size: 14, weight: .bold))
                            .foregroundColor(theme.textPrimary)
                            .lineLimit(1)
                        Spacer()
                        Button(action: { handleRemove(file) }) {
                            Text("×")
                                .font(.system(size: 18, weight: .bold))
                                .foregroundColor(theme.textSecondary)
                                .padding(8)
                        }
                        .buttonStyle(.plain)
                        .disabled(disabled)
                        .accessibilityLabel("Remove \(file.name)")
                    }
                    .padding(.leading, 12)
                    .padding(.vertical, 2)
                    .background(theme.surfaceRaised)
                    .overlay(RoundedRectangle(cornerRadius: 14).stroke(theme.borderDefault, lineWidth: 1))
                    .clipShape(RoundedRectangle(cornerRadius: 14))
                }
            }
        }
        .opacity(disabled ? 0.45 : 1)
    }

    private func handleAdd() {
        guard !disabled else { return }
        let picked: [TspUploadFile]
        if let onAdd {
            picked = onAdd()
        } else {
            let stamp = String(Int(Date().timeIntervalSince1970 * 1000))
            let name: String
            if let accept, !accept.isEmpty {
                name = "file.\(accept)"
            } else {
                name = "mock-file.txt"
            }
            picked = [TspUploadFile(id: stamp, name: name)]
        }
        guard !picked.isEmpty else { return }
        onChange(multiple ? files + picked : Array(picked.prefix(1)))
    }

    private func handleRemove(_ file: TspUploadFile) {
        guard !disabled else { return }
        onRemove(file)
        onChange(files.filter { $0.id != file.id })
    }
}

File: flutter/library/lib/src/tsp_upload.dart
import 'package:flutter/material.dart';
import 'star_planet_theme.dart';

/// Lightweight file list + add trigger (no real upload backend; add via [onAdd] mock).
class TspUploadFile {
  const TspUploadFile({required this.id, required this.name});
  final String id;
  final String name;

  factory TspUploadFile.fromDynamic(dynamic raw, int index) {
    if (raw is TspUploadFile) return raw;
    if (raw is String) return TspUploadFile(id: '$index', name: raw);
    if (raw is Map) {
      return TspUploadFile(
        id: '${raw['id'] ?? index}',
        name: '${raw['name'] ?? 'file'}',
      );
    }
    return TspUploadFile(id: '$index', name: '$raw');
  }
}

class TspUpload extends StatelessWidget {
  const TspUpload({
    super.key,
    this.files = const [],
    this.multiple = true,
    this.disabled = false,
    this.accept,
    this.theme = StarPlanetTheme.sky,
    this.onChanged,
    this.onRemove,
    this.onAdd,
  });

  final List<dynamic> files;
  final bool multiple;
  final bool disabled;
  final String? accept;
  final StarPlanetTheme theme;
  final ValueChanged<List<TspUploadFile>>? onChanged;
  final ValueChanged<TspUploadFile>? onRemove;

  /// Optional mock picker. When null, taps add a placeholder file.
  final Future<List<TspUploadFile>> Function()? onAdd;

  List<TspUploadFile> get _list =>
      files.asMap().entries.map((e) => TspUploadFile.fromDynamic(e.value, e.key)).toList();

  Future<void> _handleAdd() async {
    if (disabled) return;
    List<TspUploadFile> picked;
    if (onAdd != null) {
      picked = await onAdd!();
    } else {
      final stamp = DateTime.now().millisecondsSinceEpoch;
      picked = [TspUploadFile(id: '$stamp', name: accept?.isNotEmpty == true ? 'file.$accept' : 'mock-file.txt')];
    }
    if (picked.isEmpty) return;
    final next = multiple ? [..._list, ...picked] : picked.take(1).toList();
    onChanged?.call(next);
  }

  void _handleRemove(TspUploadFile file) {
    if (disabled) return;
    onRemove?.call(file);
    onChanged?.call(_list.where((f) => f.id != file.id).toList());
  }

  @override
  Widget build(BuildContext context) {
    final list = _list;
    return Opacity(
      opacity: disabled ? .45 : 1,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          GestureDetector(
            onTap: disabled ? null : _handleAdd,
            child: Container(
              padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 12),
              decoration: BoxDecoration(
                color: theme.brandSubtle,
                borderRadius: BorderRadius.circular(16),
                border: Border.all(color: theme.borderDefault, width: 2),
              ),
              child: Text(
                '选择文件',
                textAlign: TextAlign.center,
                style: TextStyle(color: theme.brandPrimary, fontWeight: FontWeight.w800, fontSize: 14),
              ),
            ),
          ),
          const SizedBox(height: 10),
          if (list.isEmpty)
            Text(
              '尚未选择文件',
              style: TextStyle(color: theme.textTertiary, fontWeight: FontWeight.w600, fontSize: 13),
            )
          else
            ...list.map(
              (file) => Padding(
                padding: const EdgeInsets.only(bottom: 8),
                child: Container(
                  padding: const EdgeInsets.fromLTRB(12, 10, 6, 10),
                  decoration: BoxDecoration(
                    color: theme.surfaceRaised,
                    borderRadius: BorderRadius.circular(14),
                    border: Border.all(color: theme.borderDefault),
                  ),
                  child: Row(
                    children: [
                      Expanded(
                        child: Text(
                          file.name,
                          style: TextStyle(color: theme.textPrimary, fontWeight: FontWeight.w700, fontSize: 14),
                          overflow: TextOverflow.ellipsis,
                        ),
                      ),
                      GestureDetector(
                        onTap: disabled ? null : () => _handleRemove(file),
                        behavior: HitTestBehavior.opaque,
                        child: Padding(
                          padding: const EdgeInsets.all(8),
                          child: Text(
                            '×',
                            style: TextStyle(color: theme.textSecondary, fontSize: 18, fontWeight: FontWeight.w700),
                          ),
                        ),
                      ),
                    ],
                  ),
                ),
              ),
            ),
        ],
      ),
    );
  }
}
